### ISSUE TITLE
PipelineStage issues

### ISSUE THREAD

**Issue by wconstab** [pytorch/PiPPy#941]
Just dumping issues here as I find them (applying PipelineStage to torchtrain)

1. fwd_outputs all forced to have 'requires_grad=True' -- why? what's our design here? `freqs_cis` could be passed from stage0 to stage1 but is an input value from dataloader and should not require grads
2. complex numbers aren't supported by send/recv, but `freqs_cis` is a complex tensor.  We can work around this in the model design by avoiding sending this value, but should PPStage support this automatically?
3. manually computing input sizes for each stage is tedious.  doing shape inference inside __init__ is also tedious since the 'inputs_meta' you pass in has to match the meta-ness (or cpu/cuda-ness) of the model you pass in.  Can we just do shape inference lazily on first forward call?
4. dynamic shapes- what if seqlen changes per batch? 
5. Loss fn is not yet supported

**Comment by H-Huang** [comment]
Thanks for the compiling this, I also hit a lot of these issues mentioned and Ke and I had some ideas so we can discuss it and update this issue. Also some issues of mine:

Stage
1. need to add mesh / process group support to simplify stage constructor. Stages need to be updated away from recv from rank-1 and send to rank+1 and instead use pp dimension.
2. improve erroring on `fwd()`, usually from logs we can tell which rank threw the error, but with virtual pipeline stages it is hard to tell which stage_id / model chunk threw the error.

Schedule
1. microbatches is currently expected in the form [mb_0, mb_1, mb_2, ... mb_n] where one microbatch looks like (arg_0, arg_1, arg_2, ... ,arg_n). Is there a way to make this less clunky and error prone?
2. data loading is only needed for stage 0 of the pipeline, rest of the stages dont need pass in the microbatches.
3. some schedules take list of stages and some take a single stage, consolidate this.

**Comment by H-Huang** [comment]
I've compiled these into a few concrete work items which could resolve multiple issues:

1. https://github.com/pytorch/PiPPy/issues/945
2. https://github.com/pytorch/PiPPy/issues/946
3. https://github.com/pytorch/PiPPy/issues/947
4. https://github.com/pytorch/PiPPy/issues/949
5. https://github.com/pytorch/PiPPy/issues/948
6. https://github.com/pytorch/PiPPy/issues/950
